hcl2template/types.packer_config.go

```diff
@@ -639,6 +642,59 @@ func (cfg *PackerConfig) getCoreBuildPostProcessors(source SourceUseBlock, block
 	return res, diags
 }
 
+// GetHCPPackerRegistryBlock return the HCP registry configuration block
+// that can should be used for the current build. Right now, it should
+// use the block at the top level but support the block inside the first
+// build block with a deprecation diagnostic
+func (cfg *PackerConfig) GetHCPPackerRegistryBlock() (*HCPPackerRegistryBlock, hcl.Diagnostics) {
+	var block *HCPPackerRegistryBlock
+	var diags hcl.Diagnostics
+
+	multipleRegistryDiag := func(block *HCPPackerRegistryBlock) *hcl.Diagnostic {
+		return &hcl.Diagnostic{
+			Summary:  "Multiple HCP Packer registry block declaration",
+			Subject:  block.HCL2Ref.DefRange.Ptr(),
+			Severity: hcl.DiagError,
+			Detail: "Multiple " + buildHCPPackerRegistryLabel + " blocks have been found, only one can be defined " +
+				"in HCL2 templates. Starting with Packer 1.12.1, it is recommended to move it to the " +
+				"top-level configuration instead of within a build block.",
+		}
+	}
+	// We start by looking in the build blocks
+	for _, build := range cfg.Builds {
+		if build.HCPPackerRegistry != nil {
+			if block != nil {
+				// error multiple build block
+				diags = diags.Append(multipleRegistryDiag(build.HCPPackerRegistry))
+				continue
+			}
+			block = build.HCPPackerRegistry
+			diags = diags.Append(&hcl.Diagnostic{
+				Summary:  "Build block level " + buildHCPPackerRegistryLabel + " are deprecated",
+				Subject:  &block.DefRange,
+				Severity: hcl.DiagWarning,
+				Detail: "Starting with Packer 1.12.1, it is recommended to move it to the " +
+					"top-level configuration instead of within a build block.",
+			})
+		}
+	}
+
+	if diags.HasErrors() {
+		return nil, diags
+	}
+
+	if block != nil {
+		// then we look for the top level block
+		if cfg.HCPPackerRegistry != nil {
+			fmt.Println("that")
+			diags = diags.Append(multipleRegistryDiag(block))
+		}
+		return block, diags
+	}
```

Comment:
At this point, the two conditions could be merged into one I'd think. Also the `fmt.Println` is a leftover debug print I assume?

```suggestion
	if block != nil && cfg.HCPPackerRegistry != nil {
			diags = diags.Append(multipleRegistryDiag(block))
	}
```